### PR TITLE
Some win32 fixes

### PR DESCRIPTION
**Pull request by superm1**
These are untested, but should be at least a few small improvements for usability (if they work).

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
Here is what it looks like with the artifact from CI.

image

Still need a little bit of work for LVFS refreshes to work.
image

**Comment by hughsie** [comment]
Got any ideas on the UTF-8 stuff? I remember fixing this a long time ago with 429f72b2149d02289633b823e0e6083651865018 but this might be something else.

**Comment by superm1** [comment]
> Got any ideas on the UTF-8 stuff? I remember fixing this a long time ago with [429f72b](https://github.com/fwupd/fwupd/commit/429f72b2149d02289633b823e0e6083651865018) but this might be something else.

Not really sure

**Comment by superm1** [comment]
@hughsie here is the current status:
[output.txt](https://github.com/fwupd/fwupd/files/7714414/output.txt)
The sha1 checksum problem I'm baffled by, maybe you have some ideas.

Please review all the changes and if it looks good I say merge with where this is now and sort out the checksum thing separately.

**Comment by hughsie** [comment]
>  The sha1 checksum problem I'm baffled by, maybe you have some ideas.

https://cdn.fwupd.org/downloads\firmware-02681-stable.xml.gz

The slash is the wrong way around, you're probably saving a 404 HTML page.

**Comment by hughsie** [comment]
i.e. you probably want:

    -g_build_filename(foo, bar, NULL);
    +g_build_path("/", foo, bar, NULL);

**Comment by superm1** [comment]
> > The sha1 checksum problem I'm baffled by, maybe you have some ideas.
> 
> [https://cdn.fwupd.org/downloads\firmware-02681-stable.xml.gz](https://cdn.fwupd.org/downloads%5Cfirmware-02681-stable.xml.gz)
> 
> The slash is the wrong way around, you're probably saving a 404 HTML page.



> The slash is the wrong way around, you're probably saving a 404 HTML page.

No, I checked the file using `sha1sum` in WSL and it's correctly downloaded, so that's not it.
```
$ sha1sum ~/mlimonci/AppData/Local/Microsoft/Windows/INetCache/fwupd/firmware-02681-stable.xml.gz
85e3c8d2331ba516254a85f451ed038dd5379651  /home/supermario/mlimonci/AppData/Local/Microsoft/Windows/INetCache/fwupd/firmware-02681-stable.xml.gz
```

**Comment by superm1** [comment]
> i.e. you probably want:
> 
> ```
> -g_build_filename(foo, bar, NULL);
> +g_build_path("/", foo, bar, NULL);
> ```

Great find it turns out you were right!

```
C:\Users\mlimonci>"C:\Program Files (x86)\fwupd\bin\fwupdtool.exe" refesh --verbose
<snip>
20:19:46:0449 FuCommon             running 'curl.exe https://cdn.fwupd.org/downloads/firmware.xml.gz.jcat --output C:\Users\mlimonci\AppData\Local\Microsoft\Windows\INetCache\fwupd\firmware.xml.gz.jcat --cacert C:\Program Files (x86)\fwupd\bin\..\bin\ca-bundle.crt'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2224  100  2224    0     0   2224      0  0:00:01 --:--:--  0:00:01  5270
20:19:47:0152 FuCommon             reading C:\Users\mlimonci\AppData\Local\Microsoft\Windows\INetCache\fwupd\firmware.xml.gz.jcat with 2224 bytes
20:19:47:0154 Fwupd                changing metadata URI from https://cdn.fwupd.org/downloads/firmware.xml.gz to https://cdn.fwupd.org/downloads/firmware-02681-stable.xml.gz
20:19:47:0154 FuCommon             running 'curl.exe https://cdn.fwupd.org/downloads/firmware-02681-stable.xml.gz --output C:\Users\mlimonci\AppData\Local\Microsoft\Windows\INetCache\fwupd\firmware-02681-stable.xml.gz --cacert C:\Program Files (x86)\fwupd\bin\..\bin\ca-bundle.crt'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  787k  100  787k    0     0   787k      0  0:00:01 --:--:--  0:00:01 1528k
20:19:47:0841 FuCommon             reading C:\Users\mlimonci\AppData\Local\Microsoft\Windows\INetCache\fwupd\firmware-02681-stable.xml.gz with 806314 bytes
20:19:47:0842 FuMain               updating lvfs
20:19:47:0847 Jcat                 Jcat engine kind 'gpg' not supported
20:19:47:0847 Jcat                 got 1 PKCS7 signatures
20:19:47:0849 Jcat                 verified: JcatResult:
  Timestamp:             2021-12-14T13:49:55Z
  Authority:             O=Linux Vendor Firmware Project,CN=LVFS CA
  JcatPkcs7Engine:
    Kind:                pkcs7
    VerifyKind:          signature

20:19:47:0850 Jcat                 Jcat engine kind 'gpg' not supported
20:19:47:0850 FuEngine             no existing valid keyrings: Failed to open file "C:\Users\mlimonci\fwupd\var\lib\fwupd\metadata\lvfs\metadata.xml.gz": No such file or directory
20:19:47:0851 FuCommon             writing C:\Users\mlimonci\fwupd\var\lib\fwupd\metadata\lvfs\metadata.xml.gz with 806314 bytes
20:19:47:0854 FuCommon             writing C:\Users\mlimonci\fwupd\var\lib\fwupd\metadata\lvfs\metadata.xml.gz.jcat with 2224 bytes
20:19:47:0857 XbSilo               attempting to load C:\Users\mlimonci\fwupd\var\cache\fwupd\metadata.xmlb
20:19:47:0857 XbSilo               file: 00000000-0000-0000-0000-000000000000, current:e65eeb0f-eb6f-5e46-1cd4-fe26d2df0505, cached: (NULL)
20:19:48:0251 FuEngine             1205 components now in silo
20:19:48:0263 FuEngine             requirement 1.4.5 ge 1.7.4 -> org.freedesktop.fwupd passed
```

I can check now for updates available too.
```
C:\Users\mlimonci>"C:\Program Files (x86)\fwupd\bin\fwupdtool.exe" get-updates
20:20:33:0723 FuContext            Failed to load SMBIOS: neither SMBIOS or DT found
LoadingΓÇª                 [-                                      ]20:20:33:0940 FuEngine             failed to add device usb:02:00:04: device recovery not possible
20:20:34:0019 FuEngine             failed to probe device usb:01:00:01:03:03:02: USB error on device 046d:0843 : Entity not found [-5]
LoadingΓÇª                 [***************************************]
WARNING: This package has not been validated, it may not work properly.
Devices with no available firmware updates:
 ΓÇó Jabra SPEAK 510 USB
Devices with the latest available firmware version:
 ΓÇó USB-C Dock G5
No updates available for remaining devices
```

**Comment by superm1** [comment]
Fedora 35 bump didn't fix the unicode issue, but nonethelss continues to work.  I think we should take it to get the other version deps bump anyway though.

File: libfwupdplugin/fu-common.c [hunk 1]
```diff
@@ -1224,6 +1226,28 @@ fu_common_error_array_get_best(GPtrArray *errors)
...
+ *
+ * Gets the base directory that fwupd has been launched from on Windows.
+ * This is the directory containing all subdirectories (IE 'C:\Program Files (x86)\fwupd\')
+ *
+ * Returns: The system path, or %NULL if invalid
+ *
+ * Since: 1.7.4
+ **/
+gchar *
+fu_common_get_win32_basedir(void)
```
[the remark this answers is not in the document]
Comment (superm1) [reply libfwupdplugin/fu-common.c:1240]: Yeah I think that's probably workable.

[the remark this answers is not in the document]
Comment (superm1) [reply libfwupdplugin/fu-common.c:1240]: fixed

**Comment by superm1** [comment]
with just half the hack still not right.
```
â",
â"oâ"?Jabra SPEAK 510 USB:
â",     Device ID:          b7d3afa8c7b6bf6cbd34dec25ac4e7089c16c30f
â",     Current version:    2.31
â",     Vendor:             USB:0x0B0E
â",     Serial Number:      305075317311021F00
â",     GUIDs:              443b9b32-7603-5c3a-bb30-291a7d8d6dbd
â",                         aa77a000-4286-5e17-a9cc-c62301e548e0 â+? USB\VID_0B0E&PID_0420&REV_0231
â",                         6695abef-6c60-5cbb-b8c1-2228b2c389e5 â+? USB\VID_0B0E&PID_0420
â",     Device Flags:       â?¢ Updatable
â",
```

**Review by hughsie — APPROVED** [review]

**Pull request merged**